README.md
# ATM
https://drive.google.com/drive/u/2/recent
THIS IS DEMO VIDEO FUNCTION.


History.txt
Tai khoan 20021461 da chuyen 10000 dong cho tai khoan 20021461!
Tai khoan 20021461 da duoc nap 50000 dong!
Tai khoan 20021461 da rut 10000 dong!
Tai khoan 20021461 da duoc nap 10000 dong!
Tai khoan 20021461 da rut 500000 dong!
Tai khoan 20021461 da chuyen 500000 dong cho tai khoan 20021462!
Tai khoan 20021462 da rut 965000 dong!
Tai khoan 20022002 da duoc nap 800000 dong!
Tai khoan 20022002 da rut 50000 dong!
Tai khoan 20022002 da rut 70000 dong!
Tai khoan 20022002 da chuyen 30000 dong cho tai khoan 20021461!
Tai khoan 20021234 da duoc nap 2500000 dong!
Tai khoan 20021234 da rut 10000 dong!
Tai khoan 20021234 da rut 400000 dong!
Tai khoan 20021234 da chuyen 200000 dong cho tai khoan 20021461!
Tai khoan 20021234 da duoc nap 2500000 dong!
Tai khoan 20021234 da rut 10000 dong!
Tai khoan 20021234 da chuyen 200000 dong cho tai khoan 20021461!
Tai khoan 20022002 da duoc nap 2500000 dong!
Tai khoan 20022002 da rut 10000 dong!
Tai khoan 20022002 da rut 500000 dong!
Tai khoan 20022002 da rut 700000 dong!
Tai khoan 20022002 da chuyen 400000 dong cho tai khoan 20021461!
Tai khoan 20021461 da rut 550000 dong!
Tai khoan 20021461 da duoc nap 500000 dong!
Tai khoan 20021461 da duoc nap 150000 dong!
Tai khoan 20021461 da rut 50000 dong!
Tai khoan 20021461 da duoc nap 1000000 dong!


System.cpp
#include "System.h"

using namespace std;
Screen::Screen(){}
void Screen::StartScreen()
{
    Account z;
    ATM_function a;
    ATM_box b;
    int x;
	InputKey y;
	here:
	system("cls");
	cout<<"                +++++++++++++++++++++++++++++++++++++++++++++" <<endl<<endl;
	cout<<"                 ____________________________________________"<<endl;
	cout<<"                |                 Mo ta                     |"<<endl;
	cout<<"                |___________________________________________|"<<endl;
	cout<<"                |                                           |"<<endl;
	cout<<"                |     De bat dau giao dich hay dang nhap    |"<<endl;
	cout<<"                |     Neu ban da co tai khoan               |"<<endl;
	cout<<"                |     Bam 1 de dang nhap!                   |"<<endl;
	cout<<"                |                                           |"<<endl;
	cout<<"                |                                           |"<<endl;
	cout<<"                |     Neu ban van chua co tai khoan.        |"<<endl;
	cout<<"                |     Bam 2 de dang ky!!                    |"<<endl;
	cout<<"                |                                           |"<<endl;
	cout<<"                |     De kiem tra tinh trang cay ATM        |"<<endl;
	cout<<"                |     Bam 3 !                               |"<<endl;
	cout<<"                |                                           |"<<endl;
	cout<<"                |     Neu ban da hoan thanh giao dich       |"<<endl;
	cout<<"                |     Bam 0 de thoat!                       |"<<endl;
	cout<<"                |___________________________________________|"<<endl<<endl<<endl;
	cout<<"                +++++++++++++++++++++++++++++++++++++++++++++"<<endl;
	x = y.gettheKey();
    system("cls");
	if(x == 1)
    {
        Card m;
        int n = m.Read_card();
        if(n == 3)
        {
            noti(3);
            a.logout();
        }
        if(n == 4)
        {
            noti(4);
            a.logout();
        }
	}
	if(x == 2)
    {
        z.M_register();
    }
    if(x == 3)
    {
        b.check_box();
    }
    if(x == 0)
    {
        exit(0);
    }
    else
    {
        noti(7);
        goto here;
    }
}
void Screen::menu(Account u)
{
    InputKey a;
    ATM_function y;
	cout<<"                +++++++++++++++++++++++++++++++++++++++++++++" <<endl;
	cout<<"                |Xin chao nguoi dung ^^. Hay nhap lua chon  |"<<endl;
	cout<<"                |___________________________________________|"<<endl;
	cout<<"                |                                           |"<<endl;
	cout<<"                |      1. Nap tien                          |"<<endl;
	cout<<"                |      2. Rut tien                          |"<<endl;
	cout<<"                |      3. Chuyen tien den ID                |"<<endl;
	cout<<"                |      4. Kiem tra so du                    |"<<endl;
	cout<<"                |      5. Doi ma PIN                        |"<<endl;
	cout<<"                |      6. Dang xuat                         |"<<endl;
	cout<<"                |                                           |"<<endl;
	cout<<"                |___________________________________________|"<<endl<<endl<<endl;
	cout<<"                +++++++++++++++++++++++++++++++++++++++++++++"<<endl;
	here:
    int x = a.gettheKey();
    if(x == 1)
    {
        y.deposit(u);
    }
    if(x == 2)
    {
        y.withdraw(u);
    }
    if(x == 3)
    {
        y.transfer(u);
    }
    if(x == 4)
    {
        u.see_balance();
    }
    if(x == 5)
    {
        u.change_pincode();
    }
    if(x == 6)
    {
        y.logout();
    }
    else
    {
        noti(7);
        goto here;
    }
}
void Screen::noti(int i)
{
    switch(i)
    {
        case 2:
            cout << "Ban da nhap sai pin :( hay thu lai."<<endl;
            Sleep(200);
            break;
        case 3:
            cout << "Tai khoan cua ban da bi khoa vinh vien @@" <<endl;
            Sleep(2000);
            break;
        case 4:
            cout << "Khong co du lieu ve tai khoan nay, hay dang ky!"<<endl;
            Sleep(2000);
            break;
        case 5:
            cout << "So du tai khoan cua ban khong du de thuc hien giao dich. Hay nap them tien vao tai khoan ^^"<<endl;
            Sleep(2000);
            break;
        case 6:
            cout << "ATM khong con du tien de rut. Hay quay lai sau"<<endl;
            Sleep(2000);
            break;
        case 7:
            cout << "Lenh khong hop le, hay thu lai" << endl;
            Sleep(1000);
            break;

    }
}

InputKey::InputKey(){};
int InputKey::gettheKey()
{
	int x;
	cin >> x;
	return x;
}
int InputKey::getID()
{
    int ID;
    cout <<"Vui long nhap so ID cua ban: ";
    cin >> ID;
    return ID;
}
int InputKey::getPIN()
{
    int Pin;
    cout << "Vui long nhap ma pin: ";
    cin >> Pin;
    return Pin;
}
int InputKey::choose()
{
    int option;
    cout  << "Nhap lua chon cua ban: ";
    cin >> option;
    return option;
}

Card::Card(){};
int Card::Read_card()
{
    system("cls");
	Screen a;
	InputKey b;
	int _ID, _pin;
	int check;
	Account *uzer, num_of_acc;
	int acc = num_of_acc.countacc();
	uzer = new Account[acc];
	_ID = b.getID();
	here:
	_pin = b.getPIN();
	setdata(num_of_acc,uzer);
	for (int i = 0;i < acc ;i++)
	{
		check = uzer[i].pincheck(_ID,_pin);
		if(check == 0)
		{
			if (uzer[i].get_wrongtimes() < 5)
			{
				cout << "Dang nhap thanh cong ^^.";
				Sleep(2000);
				system("cls");
				a.menu(uzer[i]);
				return 1;
			}
			else
				return 3;
		}
		if (check == 1)
		{
			if (uzer[i].get_wrongtimes()<5)
            {
                a.noti(2);
                goto here;
            }
			else
				if (uzer[i].get_wrongtimes()>=5)
					return 3;
		}
	}
	return 4;
}


Account.cpp
#include "Account.h"

using namespace std;
Account::Account(){
	numAcc = countacc();
	id = 0;
	pin = 0;
	wrongtimes = 0;
	balance = 0;
}
int Account::getcountacc()
{
    return numAcc;
}
int Account::get_PIN()
{
    return pin;
}
int Account::get_wrongtimes()
{
    return wrongtimes;
}
int Account::get_balance()
{
    return balance;
}
int Account::get_ID()
{
    return id;
}
void Account::set_balance(int i)
{
    balance = data[i];
}
void Account::set_ID(int i)
{
    id = data[i];
}
void Account::set_pin(int i)
{
    pin = data[i];
}
void Account::set_wrongtimes(int i)
{
    wrongtimes = data[i];
}
void Account::set_data(int i,int j)
{
    data[i] = j;
}
void Account::set_numAcc(int i)
{
    numAcc = i;
}
int Account::countacc(){
	ifstream file;
	file.open("Account.txt");
	int i=0;
	while(!file.eof())
	{
		file >> data[i];
		i++;
	}
	file.close();
	int numacc = i/4;
	return numacc;
}
int Account::pincheck(int ID, int PIN)
{
    if (get_PIN() == PIN && get_ID() == ID && wrongtimes >= 5)
    {
        return 1;
    }
	if (get_PIN() == PIN && get_ID() == ID && wrongtimes < 5)
    {
        fstream file;
		file.open("Account.txt");
        for (int j = 0; j< numAcc*4; j += 4)
		{
			if (data[j+1] == ID)
            {
                data[j] = 0;
                set_wrongtimes(j);
            }
		}
		for (int i = 0;i < numAcc*4; i++) file << data[i] << endl;
		file.close();
        return 0;
    }

	if (get_PIN()!= PIN && get_ID() == ID)
	{
		fstream file;
		file.open("Account.txt");
		int j = 0;
		for (j = 0; j< numAcc*4; j=j+4)
		{
			if (data[j+1] == ID)
				break;
		}
		data[j] += 1;
		set_wrongtimes(j);
		for (int i = 0;i < numAcc*4 ;i++) file << data[i] << endl;
		file.close();
		return 1;
	}
	return 2;
}
bool Account::balance_check(int cash, Account& u)
{
    Screen a;
    ATM_box b;
    if(balance < cash)
    {
        a.noti(5);
        system("cls");
        a.menu(u);
        return false;
    }
    if(b.box[6] < cash)
    {
        a.noti(6);
        system("cls");
        a.menu(u);
        return false;
    }
    return true;
}
void Account::M_register()
{
    Screen s;
    int ID,PIN;
    fstream file;
    file.open("Account.txt");
	cout<<"                +++++++++++++++++++++++++++++++++++++++++++++" <<endl<<endl;
	cout<<"                 ____________________________________________"<<endl;
	cout<<"                |   Xin chao ^^. Xin moi dang ky tai khoan  |"<<endl;
	cout<<"                |___________________________________________|"<<endl;
	cout<<"                |                                           |"<<endl;
	cout<<"                |     Buoc 1. Nhap ID                       |"<<endl;
	cout<<"                |     Buoc 2. Nhap ma Pin                   |"<<endl;
	cout<<"                |     Buoc 3. Cho he thong kiem tra         |"<<endl;
	cout<<"                |     Buoc 4: Thuc hien giao dich           |"<<endl;
	cout<<"                |                                           |"<<endl;
	cout<<"                |___________________________________________|"<<endl<<endl<<endl;
	cout<<"                +++++++++++++++++++++++++++++++++++++++++++++"<<endl;
    here:
    system("cls");
    cout << "Nhap ID: ";
    cin >> ID;
    cout << "Nhap PIN: ";
    cin >> PIN;
    for(int i =0; i < numAcc*4; i+=4)
    {
        if(data[i+1] == ID)
        {
            cout << "Dang ky khong thanh cong, tai khoan da ton tai, hay thu nhap ID khac ^^!"<<endl;
            Sleep(2000);
            goto here;
        }
    }
    cout << "Dang ky thanh cong!!, ban co the dang nhap ngay ^^";
    Sleep(2000);
    data[numAcc*4] = 0;
    data[numAcc*4+1] = ID;
    data[numAcc*4+2] = PIN;
    data[numAcc*4+3] = 0;
    for(int i = 0;i <(numAcc+1)*4 ;i++) file << data[i] << endl;
    file.close();
    s.StartScreen();

}
void Account::change_pincode()
{
    int old,new_pin;
    Screen s;
    here:
    cout << "Nhap mat khau hien tai: ";
    cin >> old;
    if(old == pin)
    {
        int i;
        fstream file;
        file.open("Account.txt");
        for(i =0; i < numAcc*4; i++)
        {
            if(id == data[i])
                break;
        }
        cout << "Nhap mat khau moi: ";
        cin >> new_pin;
        data[i+1] = new_pin;
        set_pin(i+1);
        for(i = 0; i < numAcc*4; i++)
            file << data[i] << endl;
        file.close();
        cout << "Doi mat khau thanh cong! Hay dang nhap lai";
        s.StartScreen();
    }
    else
    {
        cout << "Ban nhap sai mat khau hien tai, hay thu lai!" << endl;
        goto here;
    }
}
void Account::see_balance()
{
    InputKey x;
    Screen s;
    cout << "So du cua tai khoan " << id << " la " << balance<< endl;
    cout << "Bam 1 de tiep tuc giao dich, bam 0 de thoat ^^!"<<endl;
    if(x.gettheKey() == 1)
    {
        system("cls");
        s.menu(*this);
    }
}
void setdata(Account &num_of_acc, Account uzer[])
{
	int j = 0;
	for (int i=0;i<num_of_acc.countacc();i++)
	{
		uzer[i].set_wrongtimes(j);
		uzer[i].set_ID(j+1);
		uzer[i].set_pin(j+2);
		uzer[i].set_balance(j+3);
		j += 4;
	}
}


Trader.cpp
#include "Trader.h"
ATM_function::ATM_function(){}
void ATM_function::deposit(Account u)
{
    InputKey x;
    Screen s;
    ATM_box a;
    fstream file;
    file.open("Account.txt");
    int i;
    int k = u.countacc();
    for(i =0; i < k*4; i++)
    {
        if(u.get_ID() == u.data[i])
            break;
    }
    int currency;
    int amount;
    cout << "Nhap menh gia: ";
    cin >> currency;
    cout << "Nhap so to: ";
    cin >> amount;
    int cash = currency*amount;
    int z = u.data[i+2] + cash;
    u.set_data(i+2, z);
    u.set_balance(i+2);
    for(int j = 0; j < k*4; j++)
        file << u.data[j] << endl;
    file.close();
    file.open("ATM_box.txt");
    if(currency == 10000)
    {
        a.box[0] += amount;
    }
    if(currency == 20000)
    {
        a.box[1] += amount;
    }
    if(currency == 50000)
    {
        a.box[2] += amount;
    }
    if(currency == 100000)
    {
        a.box[3] += amount;
    }
    if(currency == 200000)
    {
        a.box[4] += amount;
    }
    if(currency == 500000)
    {
        a.box[5] += amount;
    }
    a.update_box(a);
    for(int i =0;i < 7; i++)
        file << a.box[i] << endl;
    file.close();
    cout << "Nap tien thanh cong ^^!" << endl;
    cout << "Bam 1 de tiep tuc giao dich, bam 0 de thoat ^^!"<<endl;
    file.open("History.txt",ios::app);
    file << "Tai khoan " << u.get_ID() << " da duoc nap " << cash << " dong!" << endl;
    file.close();
    if(x.gettheKey() == 1)
    {
        system("cls");
        s.menu(u);
    }
    else logout();
}
void ATM_function::withdraw(Account u)
{
    ATM_box a;
    Screen s;
    InputKey x;
    fstream file;
    file.open("Account.txt");
    int k = u.countacc();
    int i;
    int cash;
    for(i =0; i < k*4; i++)
    {
        if(u.get_ID() == u.data[i])
            break;
    }
    cout << "Hay nhap menh gia la boi so cua 10000 ^^!" << endl;
    cout << "Nhap so tien ban muon rut: ";
    cin >> cash;
    int temp = cash;
    if(u.balance_check(cash, u) == true)
    {
        u.data[i+2] -= cash;
        u.set_balance(i+2);
        if(cash / 500000 != 0 && cash/500000 <= a.box[5] )
        {
            cout << "Cay ATM dua ra " << cash/500000 << " to 500000" << endl;
            a.box[5] -= cash/500000;
            cash = cash - (cash/500000)*500000;
        }
        if(cash / 500000 != 0 && cash/500000 > a.box[5] && a.box[5] != 0 )
        {
            cout << "Cay ATM dua ra " << a.box[5] << " to 500000" << endl;
            cash = cash - (a.box[5])*500000;
            a.box[5] = 0;
        }

        if(cash / 200000 != 0 && cash/200000 <= a.box[4])
        {
            cout << "Cay ATM dua ra " << cash/200000 << " to 200000" << endl;
            a.box[4] -= cash/200000;
            cash = cash - (cash/200000)*200000;
        }
        if(cash / 200000 != 0 && cash/200000 > a.box[4] && a.box[4] != 0)
        {
            cout << "Cay ATM dua ra " << a.box[4] << " to 500000" << endl;
            cash = cash - (a.box[4])*200000;
            a.box[4] = 0;
        }

        if(cash/100000 != 0 && cash/100000 <= a.box[3])
        {
            cout << "Cay ATM dua ra " << cash/100000 << " to 100000" << endl;
            a.box[3] -= cash/100000;
            cash = cash - (cash/100000)*100000;
        }
        if(cash/100000 != 0 && cash/100000 > a.box[3] && a.box[3] != 0)
        {
            cout << "Cay ATM dua ra " << a.box[3] << " to 100000" << endl;
            cash = cash - (a.box[3])*100000;
            a.box[3] = 0;
        }

        if(cash / 50000 != 0 && cash/50000 <= a.box[2])
        {
            cout << "Cay ATM dua ra " << cash/50000 << " to 50000" << endl;
            a.box[2] -= cash/50000;
            cash = cash - (cash/50000)*50000;
        }
        if(cash / 50000 != 0 && cash/50000 > a.box[2] && a.box[2] != 0)
        {
            cout << "Cay ATM dua ra " << a.box[2] << " to 50000" << endl;
            cash = cash - (a.box[2])*50000;
            a.box[2] = 0;
        }

        if(cash / 20000 != 0 && cash/20000 <= a.box[1])
        {
            cout << "Cay ATM dua ra " << cash/20000 << " to 20000" << endl;
            a.box[1] -= cash/20000;
            cash = cash - (cash/20000)*20000;
        }
        if(cash / 20000 != 0 && cash/20000 > a.box[1]&& a.box[1] != 0)
        {
            cout << "Cay ATM dua ra " << a.box[1] << " to 20000" << endl;
            cash = cash - (a.box[1])*20000;
            a.box[1] = 0;
        }

        if(cash / 10000 != 0 && cash/10000 <= a.box[0])
        {
            cout << "Cay ATM dua ra " << cash/10000 << " to 10000" << endl;
            a.box[1] -= cash/10000;
            cash = cash - (cash/10000)*10000;
        }
        if(cash / 10000 != 0 && cash/10000 > a.box[0] && a.box[0] != 0)
        {
            cout << "Cay ATM dua ra " << a.box[0] << " to 10000" << endl;
            cash = cash - (a.box[0])*10000;
            a.box[1] = 0;
        }
        if(cash != 0)
        {
            u.data[i+2] += cash;
            cout << "Hien tai ATM khong du tien de rut het menh gia nay! Da hoan tra " << cash << " vao tai khoan." << endl;
        }
        u.set_balance(i+2);
        a.update_box(a);
        for(int i = 0; i < k*4; i++)
            file << u.data[i] << endl;
        file.close();
        file.open("ATM_box.txt");
        for(int i =0;i < 7; i++)
            file << a.box[i] << endl;
        file.close();
        file.open("History.txt",ios::app);
        file << "Tai khoan " << u.get_ID() << " da rut " << temp - cash << " dong!" << endl;
        file.close();
        cout << "Bam 1 de tiep tuc giao dich, bam 0 de thoat ^^!"<<endl;
        if(x.gettheKey() == 1)
        {
            system("cls");
            s.menu(u);
        }
        else logout();
    }
}
void ATM_function::transfer(Account u)
{
    long amount;
    InputKey x;
    Screen s;
    fstream file;
    int ID;
    int k = u.countacc();
    int i;
    int j;
    file.open("Account.txt");
    here:
    cout << "Moi ban nhap ID nguoi nhan:";
    cin >> ID;
    for( i =0 ; i < k*4 ; i++)
    {
        if(u.get_ID() == u.data[i])
            break;
    }
    for(j = 0 ; j < k*4 ; j++)
    {
        if(u.data[j] == ID)
            break;
    }
    if(j == k*4)
    {
        cout << "ID ban nhap khong ton tai, hay kiem tra lai :("<<endl;
        goto here;
    }
    else
    {
        cout << "Nhap so tien ban muon gui:";
        cin >> amount;
        u.data[i+2] -= amount;
        u.data[j+2] += amount;
        cout << "Tai khoan " << u.get_ID() << " da chuyen tien cho tai khoan " << ID << " thanh cong ^^"<<endl;
        u.set_balance(i+2);
        for(int z = 0; z < k*4; z++)
            file << u.data[z] << endl;
        file.close();
        file.open("History.txt",ios::app);
        file << "Tai khoan " << u.get_ID() << " da chuyen " << amount << " dong cho tai khoan " << ID << "!" << endl;
        file.close();
        cout << "Bam 1 de tiep tuc giao dich, bam 0 de thoat ^^!"<<endl;
        if(x.gettheKey() == 1)
        {
            system("cls");
            s.menu(u);
        }
        else logout();
    }
}
void ATM_function::logout()
{
    Screen s;
    s.StartScreen();
}

ATM_box::ATM_box()
{
    fstream file;
    file.open("ATM_box.txt");
    for(int i =0; i < 7 ; i++)
        file >> box[i];
    m10 = box[0];
    m20 = box[1];
    m50 = box[2];
    m100 = box[3];
    m200 = box[4];
    m500 = box[5];
    box[6] = box[0]*10000+box[1]*20000+box[2]*50000+box[3]*100000+box[4]*200000+box[5]*500000;
    file.close();
}
void ATM_box::update_box(ATM_box &b)
{
    b.box[6] = b.box[0]*10000+b.box[1]*20000+b.box[2]*50000+b.box[3]*100000+b.box[4]*200000+b.box[5]*500000;
}
void ATM_box::check_box()
{
    fstream file;
    InputKey x;
    Screen s;
    file.open("ATM_box.txt");
    for(int i =0; i < 7 ; i++)
        file >> box[i];
    cout<<"                +++++++++++++++++++++++++++++++++++++++++++++" <<endl;
	cout<<"                |Xin chao ^^. Day la tinh trang cua ATM     |"<<endl;
	cout<<"                |___________________________________________|"<<endl;
	cout<<"                |                                           |"<<endl;
	cout<<"                |      1. So to menh gia 10000  :" << box[0] <<"         |"<<endl;
	cout<<"                |      2. So to menh gia 20000  :" << box[1] <<"         |"<<endl;
	cout<<"                |      3. So to menh gia 50000  :" << box[2] <<"         |"<<endl;
	cout<<"                |      4. So to menh gia 100000 :" << box[3] <<"         |"<<endl;
	cout<<"                |      5. So to menh gia 200000 :" << box[4] <<"         |"<<endl;
	cout<<"                |      6. So to menh gia 500000 :" << box[5] <<"         |"<<endl;
	cout<<"                |      7. Tong so tien trong ATM:" << box[6] <<"   |"<<endl;
	cout<<"                |                                           |"<<endl;
	cout<<"                |___________________________________________|"<<endl<<endl<<endl;
	cout<<"                +++++++++++++++++++++++++++++++++++++++++++++"<<endl<<endl<<endl;
    file.close();
    cout << "Bam 1 de dang nhap, bam phim bat ki de thoat chuong trinh ^^!"<<endl;
    if(x.gettheKey() == 1)
    {
        s.StartScreen();
    }

}
